### PR TITLE
feat: Implement mixins

### PR DESCRIPTION
**Pull request by EdVraz**
This PR aims to implement important mixnins

draft until I made the docs

### CODE REVIEW

**Comment by EdVraz** [comment]
added docs, this is ready for review @i0bs

File: retux/client/mixins/controllable.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+__all__ = ["Controllable"]
+
+
+class Controllable:
+    """
+    A mixin for objects with create/get methods.
+
+    Methods
+    _______
```
Comment (i0bs) [inline retux/client/mixins/controllable.py:13]:
```suggestion
    -------
```

File: retux/client/mixins/controllable.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        Gets an object from the Discord API.
+    """
+
+    @classmethod
+    async def create(cls, bot: Bot, path: str, **kwargs) -> dict:
+        """
+        Creates an object with the given path and kwargs.
+
+        Attributes
+        ___________
```
Comment (i0bs) [inline retux/client/mixins/controllable.py:26]:
```suggestion
        ----------
```

File: retux/client/mixins/controllable.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        bot : `retux.Bot`
+            The instance of the bot.
+        path : `str`
+            The path to the endpoint of the Discord API.
+        **kwargs : `dict`
+            The data for the object creation.
+
+        Returns
+        _______
```
Comment (i0bs) [inline retux/client/mixins/controllable.py:35]:
```suggestion
        -------
```

File: retux/client/mixins/controllable.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        `dict`
+            The data Discord returned as dict.
```
Comment (i0bs) [inline retux/client/mixins/controllable.py:37]:
```suggestion
            The data of the object returned by Discord.
```

File: retux/client/mixins/controllable.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+        return await bot.http.request(route, payload)
+
+    @classmethod
+    async def get(cls, bot: Bot, path: str, **query_params) -> dict:
+        """
+        Gets an object from the Discord API.
+
+        Attributes
+        __________
```
Comment (i0bs) [inline retux/client/mixins/controllable.py:58]:
```suggestion
        ----------
```

File: retux/client/mixins/controllable.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        bot : `retux.Bot`
+            The instance of the bot.
+        path : `str`
+            The path to the endpoint of the Discord API.
+        **query_params : `dict`
+            The data to query for when getting. For example `with_counts=True` for guilds.
+
+        Returns
+        _______
```
Comment (i0bs) [inline retux/client/mixins/controllable.py:67]:
```suggestion
        -------
```

File: retux/client/mixins/controllable.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        `dict`
+            The data Discord returned as dict.
```
Comment (i0bs) [inline retux/client/mixins/controllable.py:69]:
```suggestion
            The data of the object returned by Discord.
```

File: retux/client/mixins/editable.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+__all__ = ["Editable"]
+
+
+class Editable:
+    """
+    A mixin for objects with edit/delete methods.
+
+    Methods
+    _______
```
Comment (i0bs) [inline retux/client/mixins/editable.py:13]:
```suggestion
    -------
```

File: retux/client/mixins/editable.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    `modify()` : `dict`
+        An alias of `edit()`
+    `delete()` : `dict | None`
+        Delete an object from discord.
+    """
+
+    async def edit(self, bot: Bot, path: str, **kwargs) -> dict:
+        """
+        Attributes
+        __________
```
Comment (i0bs) [inline retux/client/mixins/editable.py:25]:
```suggestion
        ----------
```

File: retux/client/mixins/editable.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        bot : `retux.Bot`
+            The instance of the bot.
+        path : `str`
+            The path to the endpoint of the Discord API.
+        **kwargs : `dict`
+            The data to edit
+
+        Returns
+        _______
```
Comment (i0bs) [inline retux/client/mixins/editable.py:34]:
```suggestion
        -------
```

File: retux/client/mixins/editable.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+                payload[key] = asdict(
+                    value, filter=lambda _name, _value: _name.name not in {"_bot", "bot"}
+                )
+            else:
+                payload[key] = value
+
+        return await bot.http.request(route, payload)
+
+    async def modify(self, bot: Bot, path: str, **kwargs) -> dict:
+        """Alias of `edit()`"""
```
Comment (i0bs) [inline retux/client/mixins/editable.py:52]:
```suggestion
        """An alias of the `edit()` method."""
```

File: retux/client/mixins/editable.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    `edit()` : `dict`
+        Edits an object with the given path and kwargs.
+    `modify()` : `dict`
+        An alias of `edit()`
+    `delete()` : `dict | None`
+        Delete an object from discord.
+    """
+
+    async def edit(self, bot: Bot, path: str, **kwargs) -> dict:
+        """
```
Comment (i0bs) [inline retux/client/mixins/editable.py:23]: This method is documented in the class' docstring but lacks the documentation in the method itself.

```suggestion
        """
        Edits an object with the given path and kwargs.

```

File: retux/client/mixins/editable.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        Attributes
+        __________
+        bot : `retux.Bot`
+            The instance of the bot.
+        path : `str`
+            The path to the endpoint of the Discord API.
+        **kwargs : `dict`
+            The data to edit
```
Comment (i0bs) [inline retux/client/mixins/editable.py:31]:
```suggestion
            The data to edit.
```

File: retux/client/mixins/editable.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    async def modify(self, bot: Bot, path: str, **kwargs) -> dict:
+        """Alias of `edit()`"""
+        return await self.edit(bot, path, **kwargs)
+
+    async def delete(self, bot: Bot, path: str) -> None | dict:
+        """
+        Delete an object from discord.
+
+        Attributes
+        __________
```
Comment (i0bs) [inline retux/client/mixins/editable.py:60]:
```suggestion
        ----------
```

File: retux/client/mixins/editable.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+        bot : `retux.Bot`
+            The instance of the bot.
+        path : `str`
+            The path to the endpoint of the Discord API.
+
+        Returns
+        _______
```
Comment (i0bs) [inline retux/client/mixins/editable.py:67]:
```suggestion
        -------
```

File: retux/client/mixins/respondable.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+__all__ = ["Respondable"]
+
+
+class Respondable(Editable):
+    """
+    A mixin for objects with send/edit/delete methods
+
+    Methods
+    _______
```
Comment (i0bs) [inline retux/client/mixins/respondable.py:15]:
```suggestion
    -------
```

File: retux/client/mixins/respondable.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+from .editable import Editable
+
+from attrs import asdict
+
+__all__ = ["Respondable"]
+
+
+class Respondable(Editable):
+    """
+    A mixin for objects with send/edit/delete methods
```
Comment (i0bs) [inline retux/client/mixins/respondable.py:12]:
```suggestion
    A mixin for objects with send/edit/delete methods.
```

File: retux/client/mixins/respondable.py [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+    Methods
+    _______
+    respond() : `dict`
+        Executes a `respond` action to discord with the given Route and keyword arguments.
+    send() : `dict`
+        An alias of `respond()`
```
Comment (i0bs) [inline retux/client/mixins/respondable.py:19]:
```suggestion
        An alias of the `respond()` method.
```

File: retux/client/mixins/respondable.py [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+    """
+
+    async def respond(self, bot: Bot, path: str, **kwargs) -> dict:
+        """
+        Executes a `respond` action to discord with the given Route and keyword arguments.
```
Comment (i0bs) [inline retux/client/mixins/respondable.py:24]: I get what's being meant here but I'm lost what the "respond action" is. Did you mean "Executes a `Create Interaction Response` action" ?

```suggestion
        Executes a `respond` action to Discord with the given route and keyword arguments.
```

File: retux/client/mixins/respondable.py [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+
+        Attributes
+        __________
```
Comment (i0bs) [inline retux/client/mixins/respondable.py:27]:
```suggestion
        ----------
```

File: retux/client/mixins/respondable.py [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+        bot : `retux.Bot`
+            The instance of the bot.
+        path : `str`
+            The path to the endpoint of the Discord API.
+        **kwargs : `dict`
+            The data to include in the request.
+
+        Returns
+        _______
```
Comment (i0bs) [inline retux/client/mixins/respondable.py:36]:
```suggestion
        -------
```

File: retux/client/mixins/respondable.py [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+        `dict`
+            The data Discord returned as dict.
```
Comment (i0bs) [inline retux/client/mixins/respondable.py:38]:
```suggestion
            The data of the interaction response returned by Discord.
```

File: retux/client/mixins/respondable.py [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+
+
+class Respondable(Editable):
+    """
+    A mixin for objects with send/edit/delete methods
+
+    Methods
+    _______
+    respond() : `dict`
+        Executes a `respond` action to discord with the given Route and keyword arguments.
```
Comment (i0bs) [inline retux/client/mixins/respondable.py:17]: See review comment left on the `respond()` method's docstring.

File: retux/client/mixins/respondable.py [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+                payload[key] = asdict(
+                    value, filter=lambda _name, _value: _name.name not in {"_bot", "bot"}
+                )
+            else:
+                payload[key] = value
+
+        return await bot.http.request(route, payload)
+
+    async def send(self, bot: Bot, path: str, **kwargs) -> dict:
+        """Alias of `respond()`"""
```
Comment (i0bs) [inline retux/client/mixins/respondable.py:55]:
```suggestion
        """An alias of the `respond()` method."""
```

**Review by i0bs — CHANGES_REQUESTED** [review]
I don't know if PyCharm is being weird to you, but this just reflects some corrections on docstring styles.

Also, as a side note, I've realised that it may be a little inefficient for us to throw every single mixin into a separate file. It's probably great for organisation purposes, but now that I look at it, I can't imagine us building mixins to go into mixins for instance. Maybe we should move them into just a `mixins.py`?

Anywho, first revision of changes requested.

**Review by i0bs — APPROVED** [review]
LGTM, thank you for contributing to retux!

**Pull request merged**